src/psyclone/tests/psyad/tl2ad_harness_test.py

```diff
@@ -0,0 +1,486 @@
+    code = fortran_writer(nodes[1])
+    assert "result = result + var1 * var2" in code
+
+
+def test_create_inner_product_1d_arrays(fortran_writer):
+    ''' Test for utility that creates PSyIR for computing an
+    inner product when given rank-1 arrays. '''
+    accum = DataSymbol("result", REAL_DOUBLE_TYPE)
+    array_type = ArrayType(INTEGER_TYPE, [10])
+    var1 = DataSymbol("var1", array_type)
+    var2 = DataSymbol("var2", array_type)
+    nodes = _create_inner_product(accum, [(var1, var2)])
+    assert len(nodes) == 2
+    assert isinstance(nodes[0], Assignment)
+    assert nodes[0].lhs.symbol is accum
+    assert nodes[0].rhs.value == "0.0"
+    assert isinstance(nodes[1], Assignment)
+    assert nodes[1].lhs.symbol is accum
+    assert isinstance(nodes[1].rhs, BinaryOperation)
+    assert nodes[1].rhs.operator == BinaryOperation.Operator.ADD
+    code = fortran_writer(nodes[1])
+    assert "result = result + DOT_PRODUCT(var1, var2)" in code
+
+
+def test_create_inner_product_arrays(fortran_writer):
+    ''' Test for utility that creates PSyIR for computing an
+    inner product when given arrays with rank > 1. '''
+    accum = DataSymbol("result", REAL_DOUBLE_TYPE)
+    array_type = ArrayType(INTEGER_TYPE, [10, 10, 10])
+    var1 = DataSymbol("var1", array_type)
+    var2 = DataSymbol("var2", array_type)
+    nodes = _create_inner_product(accum, [(var1, var2)])
+    assert len(nodes) == 2
+    assert isinstance(nodes[0], Assignment)
+    assert nodes[0].lhs.symbol is accum
+    assert nodes[0].rhs.value == "0.0"
+    assert isinstance(nodes[1], Assignment)
+    assert nodes[1].lhs.symbol is accum
+    assert isinstance(nodes[1].rhs, BinaryOperation)
+    assert nodes[1].rhs.operator == BinaryOperation.Operator.ADD
+    code = fortran_writer(nodes[1])
+    assert "result = result + SUM(var1(:,:,:) * var2(:,:,:))" in code
+
+
+def test_inner_product_scalars_and_arrays(fortran_writer):
+    ''' Test for utility that creates PSyIR for computing an
+    inner product when given arrays and scalars. '''
+    accum = DataSymbol("result", REAL_DOUBLE_TYPE)
+    array3d_type = ArrayType(INTEGER_TYPE, [10, 10, 10])
+    vars3d = DataSymbol("var1", array3d_type), DataSymbol("var2", array3d_type)
+    array1d_type = ArrayType(INTEGER_TYPE, [5])
+    vecs = DataSymbol("vec1", array1d_type), DataSymbol("vec2", array1d_type)
+    scals = DataSymbol("a1", REAL_TYPE), DataSymbol("a2", REAL_TYPE)
+    nodes = _create_inner_product(accum, [vars3d, vecs, scals])
+    assert len(nodes) == 4
+    assert all([isinstance(node, Assignment) for node in nodes])
+    assert fortran_writer(nodes[0]) == "result = 0.0\n"
+    assert (fortran_writer(nodes[1]) ==
+            "result = result + SUM(var1(:,:,:) * var2(:,:,:))\n")
+    assert (fortran_writer(nodes[2]) ==
+            "result = result + DOT_PRODUCT(vec1, vec2)\n")
+    assert (fortran_writer(nodes[3]) == "result = result + a1 * a2\n")
+
+
+def test_generate_adjoint_test_errors():
```

Comment:
The `*adjoint_test*` tests should be before the `*inner_product*` tests as they should be provided in the same order as they are in the original file (for ease of navigation).

Reply:
OK. I don't think we have a rule about this though. Perhaps we should?

Reply:
I'll start a discussion on teams about this.